File: src/byte_vector.c
#include "byte_vector.h"
#include <stdlib.h>

#define INITIAL_CAPACITY 16
#define GROWTH_FACTOR 2

static void bv_grow(ByteVector *bv) {
  size_t new_capacity =
      bv->cap == 0 ? INITIAL_CAPACITY : bv->cap * GROWTH_FACTOR;
  uint8_t *new_data = realloc(bv->data, new_capacity);
  bv->data = new_data;
  bv->cap = new_capacity;
}

ByteVector *bv_create(void) {
  return bv_create_with_capacity(INITIAL_CAPACITY);
}

ByteVector *bv_create_with_capacity(size_t capacity) {
  ByteVector *bv = malloc(sizeof(ByteVector));
  bv->data = capacity > 0 ? malloc(capacity) : NULL;
  bv->len = 0;
  bv->cap = capacity;
  return bv;
}

void bv_destroy(ByteVector *bv) {
  free(bv->data);
  free(bv);
}

void bv_push(ByteVector *bv, uint8_t byte) {
  if (bv->len >= bv->cap) {
    bv_grow(bv);
  }
  bv->data[bv->len] = byte;
  bv->len += 1;
}

uint8_t bv_pop(ByteVector *bv) {
  uint8_t result = bv->data[bv->len];
  bv->len -= 1;
  return result;
}

uint8_t bv_get(const ByteVector *bv, size_t index) { return bv->data[index]; }

void bv_set(ByteVector *bv, size_t index, uint8_t byte) {
  bv->data[index] = byte;
}

void bv_clear(ByteVector *bv) { bv->len = 0; }

File: lib/byte_vector.dart
/// Support for doing something awesome.
///
/// More dartdocs go here.
library;

import 'dart:ffi';

import 'src/byte_vector.g.dart' as bv;

class ByteVector {
  final Pointer<bv.ByteVector> _vector;

  ByteVector() : _vector = bv.bv_create();

  int get length => _vector.ref.len;

  int get capacity => _vector.ref.cap;

  void add(int byte) {
    bv.bv_push(_vector, byte);
  }

  int operator [](int index) {
    return bv.bv_get(_vector, index);
  }

  void operator []=(int index, int value) {
    return bv.bv_set(_vector, index, value);
  }
}
